panic: "duplicate checkable objects report" on applying ephemeral resources

### Terraform Version

```shell
1.10.0-dev (0a266c88c49a9486898faaf53e9ee9b2615a81de)
```


### Terraform Configuration Files

```terraform
terraform {
  required_providers {
    random = {
      source = "hashicorp/random"
    }
  }
}

locals {
  foo = "bar"
}

ephemeral "random_password" "example" {
  length = 12

  lifecycle {
    precondition {
      condition     = local.foo == "bar"
      error_message = "noot noot"
    }
    # postcondition {
    #   condition     = self.result != "foo"
    #   error_message = "password is not foo but"
    # }
  }
}
```


### Debug Output

https://gist.github.com/radeksimko/5de91d24856bb12f4379f16f3727fc26

### Expected Behavior

`terraform apply` to not crash

### Actual Behavior

`terraform apply` crashed with the following output whenever there's `precondition` or `postcondition` in the configuration

```
No changes. Your infrastructure matches the configuration.

Terraform has compared your real infrastructure against your configuration and found no differences, so no changes are needed.

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
Please report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version, the stack trace
shown below, and any additional information which may help replicate the issue.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

panic: duplicate checkable objects report for ephemeral.random_password.example 
goroutine 161 [running]:
runtime/debug.Stack()
        /opt/homebrew/Cellar/go/1.23.2/libexec/src/runtime/debug/stack.go:26 +0x64
github.com/hashicorp/terraform/internal/logging.PanicHandler()
        /Users/radeksimko/gopath/src/github.com/hashicorp/terraform/internal/logging/panic.go:84 +0x184
panic({0x1036cd400?, 0x14000cc7030?})
        /opt/homebrew/Cellar/go/1.23.2/libexec/src/runtime/panic.go:785 +0x124
github.com/hashicorp/terraform/internal/terraform.(*Graph).walk.func1.1()
        /Users/radeksimko/gopath/src/github.com/hashicorp/terraform/internal/terraform/graph.go:59 +0x36c
panic({0x1036cd400?, 0x14000cc7030?})
        /opt/homebrew/Cellar/go/1.23.2/libexec/src/runtime/panic.go:785 +0x124
github.com/hashicorp/terraform/internal/checks.(*State).ReportCheckableObjects(0x103c6aa20?, {0x103e15070, 0x1400052de00}, 0x14000ccabd0)
        /Users/radeksimko/gopath/src/github.com/hashicorp/terraform/internal/checks/state_report.go:33 +0x4c8
github.com/hashicorp/terraform/internal/terraform.(*nodeExpandPlannableResource).dynamicExpandPartial(0x14000cb64b0, {0x103e38958, 0x14000991e00}, {0x1059ad2a0, 0x1, 0x1}, 0x0, {0x14000b7b5b0?}, {0x14000b7b548?})
        /Users/radeksimko/gopath/src/github.com/hashicorp/terraform/internal/terraform/node_resource_partial_plan.go:210 +0x105c
github.com/hashicorp/terraform/internal/terraform.(*nodeExpandPlannableResource).DynamicExpand(0x14000cb64b0, {0x103e38958, 0x14000991e00})
        /Users/radeksimko/gopath/src/github.com/hashicorp/terraform/internal/terraform/node_resource_plan.go:128 +0x378
github.com/hashicorp/terraform/internal/terraform.(*nodeExpandApplyableResource).DynamicExpand(0x14000ca1100, {0x103e38958, 0x14000991e00})
        /Users/radeksimko/gopath/src/github.com/hashicorp/terraform/internal/terraform/node_resource_apply.go:61 +0xc0
github.com/hashicorp/terraform/internal/terraform.(*Graph).walk.func1({0x103cfd060, 0x14000ca1100})
        /Users/radeksimko/gopath/src/github.com/hashicorp/terraform/internal/terraform/graph.go:153 +0x77c
github.com/hashicorp/terraform/internal/dag.(*Walker).walkVertex(0x14000a5b800, {0x103cfd060, 0x14000ca1100}, 0x1400052d540)
        /Users/radeksimko/gopath/src/github.com/hashicorp/terraform/internal/dag/walk.go:384 +0x2a0
created by github.com/hashicorp/terraform/internal/dag.(*Walker).Update in goroutine 155
        /Users/radeksimko/gopath/src/github.com/hashicorp/terraform/internal/dag/walk.go:307 +0xc18
```

### Steps to Reproduce

1. `terraform init`
2. `terraform apply`


### Additional Context

For `hashicorp/random` provider I used a custom build from [this branch](https://github.com/hashicorp/terraform-provider-random/compare/main...av/ephemeral-password).



### References

_No response_

--- a/internal/providers/testing/provider_mock.go
+++ b/internal/providers/testing/provider_mock.go
@@ -632,11 +632,6 @@ func (p *MockProvider) RenewEphemeralResource(r providers.RenewEphemeralResource
 		return resp
 	}
 
-	if p.CloseEphemeralResourceCalled {
-		resp.Diagnostics = resp.Diagnostics.Append(fmt.Errorf("CloseEphemeralResource called on %q before RenewEphemeralResource", r.TypeName))
-		return resp
-	}
-
 	p.RenewEphemeralResourceCalled = true
 	p.RenewEphemeralResourceRequest = r
 

--- a/internal/resources/ephemeral/ephemeral_resources.go
+++ b/internal/resources/ephemeral/ephemeral_resources.go
@@ -91,7 +91,14 @@ func (r *Resources) InstanceValue(addr addrs.AbsResourceInstance) (val cty.Value
 	}
 	inst, ok := insts.GetOk(addr)
 	if !ok {
-		return cty.DynamicVal, false
+		// Here we can assume that if the entire resource exists, the instance
+		// is valid because Close removes resources as a whole. Individual
+		// instances may not actually be present when checks are evaluated,
+		// because they are evaluated from instance nodes that are using "self".
+		// The way an instance gets "self" is to call GetResource which needs to
+		// compile all instances into a suitable value, so we may be missing
+		// instances which have not yet been opened.
+		return cty.DynamicVal, true
 	}
 	// If renewal has failed then we can't assume that the object is still
 	// live, but we can still return the original value regardless.

--- a/internal/terraform/context_apply_ephemeral_test.go
+++ b/internal/terraform/context_apply_ephemeral_test.go
@@ -285,6 +285,7 @@ variable "input" {
 }
 
 ephemeral "ephem_resource" "data" {
+  for_each = toset(["a", "b"])
   lifecycle {
     precondition {
       condition = var.input == "ok"
@@ -298,7 +299,7 @@ ephemeral "ephem_resource" "data" {
 }
 
 provider "test" {
-  test_string = ephemeral.ephem_resource.data.value
+  test_string = ephemeral.ephem_resource.data["a"].value
 }
 
 resource "test_object" "test" {

--- a/internal/terraform/context_apply_ephemeral_test.go
+++ b/internal/terraform/context_apply_ephemeral_test.go
@@ -353,6 +353,9 @@ resource "test_object" "test" {
 	})
 	assertNoDiagnostics(t, diags)
 
+	// reset the ephemeral call flags
+	ephem.ConfigureProviderCalled = false
+
 	_, diags = ctx.Apply(plan, m, nil)
 	assertNoDiagnostics(t, diags)
 }

--- a/internal/terraform/context_apply_ephemeral_test.go
+++ b/internal/terraform/context_apply_ephemeral_test.go
@@ -275,3 +275,83 @@ output "data" {
 		t.Error("CloseEphemeralResourceCalled not called")
 	}
 }
+
+func TestContext2Apply_ephemeralChecks(t *testing.T) {
+	// test the full validate-plan-apply lifecycle for ephemeral conditions
+	m := testModuleInline(t, map[string]string{
+		"main.tf": `
+variable "input" {
+  type = string
+}
+
+ephemeral "ephem_resource" "data" {
+  lifecycle {
+    precondition {
+      condition = var.input == "ok"
+      error_message = "input not ok"
+    }
+    postcondition {
+      condition = self.value != null
+      error_message = "value is null"
+    }
+  }
+}
+
+provider "test" {
+  test_string = ephemeral.ephem_resource.data.value
+}
+
+resource "test_object" "test" {
+}
+`,
+	})
+
+	ephem := &testing_provider.MockProvider{
+		GetProviderSchemaResponse: &providers.GetProviderSchemaResponse{
+			EphemeralResourceTypes: map[string]providers.Schema{
+				"ephem_resource": {
+					Block: &configschema.Block{
+						Attributes: map[string]*configschema.Attribute{
+							"value": {
+								Type:     cty.String,
+								Computed: true,
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	ephem.OpenEphemeralResourceFn = func(providers.OpenEphemeralResourceRequest) (resp providers.OpenEphemeralResourceResponse) {
+		resp.Result = cty.ObjectVal(map[string]cty.Value{
+			"value": cty.StringVal("test string"),
+		})
+		return resp
+	}
+
+	p := simpleMockProvider()
+
+	ctx := testContext2(t, &ContextOpts{
+		Providers: map[addrs.Provider]providers.Factory{
+			addrs.NewDefaultProvider("ephem"): testProviderFuncFixed(ephem),
+			addrs.NewDefaultProvider("test"):  testProviderFuncFixed(p),
+		},
+	})
+
+	diags := ctx.Validate(m, &ValidateOpts{})
+	assertNoDiagnostics(t, diags)
+
+	plan, diags := ctx.Plan(m, nil, &PlanOpts{
+		SetVariables: InputValues{
+			"input": &InputValue{
+				Value:      cty.StringVal("ok"),
+				SourceType: ValueFromConfig,
+			},
+		},
+	})
+	assertNoDiagnostics(t, diags)
+
+	_, diags = ctx.Apply(plan, m, nil)
+	assertNoDiagnostics(t, diags)
+}

--- a/internal/terraform/node_resource_apply.go
+++ b/internal/terraform/node_resource_apply.go
@@ -5,6 +5,7 @@ package terraform
 
 import (
 	"github.com/hashicorp/terraform/internal/addrs"
+	"github.com/hashicorp/terraform/internal/dag"
 	"github.com/hashicorp/terraform/internal/tfdiags"
 )
 
@@ -51,14 +52,7 @@ func (n *nodeExpandApplyableResource) Name() string {
 
 func (n *nodeExpandApplyableResource) DynamicExpand(ctx EvalContext) (*Graph, tfdiags.Diagnostics) {
 	if n.Addr.Resource.Mode == addrs.EphemeralResourceMode {
-		// We need to expand the ephemeral resources the same as we do during
-		// planning, so we convert this into the plannable node on the fly.
-		// There doesn't seem to be any better way to handle this for now, since
-		// ephemeral resources need everything to happen the same as it would
-		// during planning.
-		return (&nodeExpandPlannableResource{
-			NodeAbstractResource: n.NodeAbstractResource,
-		}).DynamicExpand(ctx)
+		return n.dynamicExpandEphemeral(ctx)
 	}
 
 	var diags tfdiags.Diagnostics
@@ -70,4 +64,101 @@ func (n *nodeExpandApplyableResource) DynamicExpand(ctx EvalContext) (*Graph, tf
 	}
 
 	return nil, diags
+}
+
+// We need to expand the ephemeral resources mostly the same as we do during
+// planning. There a lot of options than happen during planning which aren't
+// applicable to apply however, and we have to make sure we don't re-register
+// checks which already recorded in the plan, so we create a pared down version
+// of the plan expansion here.
+func (n *nodeExpandApplyableResource) dynamicExpandEphemeral(ctx EvalContext) (*Graph, tfdiags.Diagnostics) {
+	var diags tfdiags.Diagnostics
+	var g Graph
+
+	expander := ctx.InstanceExpander()
+	moduleInstances := expander.ExpandModule(n.Addr.Module, false)
+
+	for _, module := range moduleInstances {
+		resAddr := n.Addr.Resource.Absolute(module)
+		expDiags := n.expandEphemeralResourceInstances(ctx, resAddr, &g)
+		diags = diags.Append(expDiags)
+	}
+
+	return &g, diags
+}
+
+func (n *nodeExpandApplyableResource) expandEphemeralResourceInstances(globalCtx EvalContext, resAddr addrs.AbsResource, g *Graph) tfdiags.Diagnostics {
+	var diags tfdiags.Diagnostics
+
+	// The rest of our work here needs to know which module instance it's
+	// working in, so that it can evaluate expressions in the appropriate scope.
+	moduleCtx := evalContextForModuleInstance(globalCtx, resAddr.Module)
+
+	// writeResourceState is responsible for informing the expander of what
+	// repetition mode this resource has, which allows expander.ExpandResource
+	// to work below.
+	moreDiags := n.recordResourceData(moduleCtx, resAddr)
+	diags = diags.Append(moreDiags)
+	if moreDiags.HasErrors() {
+		return diags
+	}
+
+	expander := moduleCtx.InstanceExpander()
+	instanceAddrs := expander.ExpandResource(resAddr)
+
+	instG, instDiags := n.ephemeralResourceInstanceSubgraph(resAddr, instanceAddrs)
+	if instDiags.HasErrors() {
+		diags = diags.Append(instDiags)
+		return diags
+	}
+	g.Subsume(&instG.AcyclicGraph.Graph)
+	return diags
+}
+
+func (n *nodeExpandApplyableResource) ephemeralResourceInstanceSubgraph(addr addrs.AbsResource, instanceAddrs []addrs.AbsResourceInstance) (*Graph, tfdiags.Diagnostics) {
+	var diags tfdiags.Diagnostics
+
+	concreteEphemeral := func(a *NodeAbstractResourceInstance) dag.Vertex {
+		a.Config = n.Config
+		a.ResolvedProvider = n.ResolvedProvider
+		a.Schema = n.Schema
+		a.ProvisionerSchemas = n.ProvisionerSchemas
+		a.ProviderMetas = n.ProviderMetas
+		a.dependsOn = n.dependsOn
+
+		// we still need the Plannable resource instance
+		return &NodeApplyableResourceInstance{
+			NodeAbstractResourceInstance: a,
+		}
+	}
+
+	// Start creating the steps
+	steps := []GraphTransformer{
+		// Expand the count or for_each (if present)
+		&ResourceCountTransformer{
+			Concrete:      concreteEphemeral,
+			Schema:        n.Schema,
+			Addr:          n.ResourceAddr(),
+			InstanceAddrs: instanceAddrs,
+		},
+
+		// Targeting
+		&TargetsTransformer{Targets: n.Targets},
+
+		// Connect references so ordering is correct
+		&ReferenceTransformer{},
+
+		// Make sure there is a single root
+		&RootTransformer{},
+	}
+
+	// Build the graph
+	b := &BasicGraphBuilder{
+		Steps: steps,
+		Name:  "nodeExpandApplyEphemeralResource",
+	}
+	graph, graphDiags := b.Build(addr.Module)
+	diags = diags.Append(graphDiags)
+
+	return graph, diags
 }

--- a/internal/terraform/node_resource_apply_instance.go
+++ b/internal/terraform/node_resource_apply_instance.go
@@ -145,9 +145,21 @@ func (n *NodeApplyableResourceInstance) Execute(ctx EvalContext, op walkOperatio
 		return n.managedResourceExecute(ctx)
 	case addrs.DataResourceMode:
 		return n.dataResourceExecute(ctx)
+	case addrs.EphemeralResourceMode:
+		return n.ephemeralResourceExecute(ctx)
 	default:
 		panic(fmt.Errorf("unsupported resource mode %s", n.Config.Mode))
 	}
+}
+
+func (n *NodeApplyableResourceInstance) ephemeralResourceExecute(ctx EvalContext) tfdiags.Diagnostics {
+	_, diags := ephemeralResourceOpen(ctx, ephemeralResourceInput{
+		addr:           n.Addr,
+		config:         n.Config,
+		providerConfig: n.ResolvedProvider,
+	})
+
+	return diags
 }
 
 func (n *NodeApplyableResourceInstance) dataResourceExecute(ctx EvalContext) (diags tfdiags.Diagnostics) {

